### PR TITLE
Player v4 update

### PR DESCRIPTION
**Pull request by marcosmx**
This branch contains the effort to migrate the chromecast receiver app from V3 to V4.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,41 +1,79 @@
...
 - For iOS, view the sample Chromecast Sender [iOS App](https://github.com/ooyala/ios-sample-apps/tree/master/ChromecastSampleApp)
 - For Android, view the sample Chromecast Sender [Android App](https://github.com/ooyala/android-sample-apps/tree/master/ChromecastSampleApp)
 - Purchase a Chromecast device
 
-See [http://support.ooyala.com/developers/documentation/concepts/chromecast_prerequisites.html](http://support.ooyala.com/developers/documentation/concepts/chromecast_prerequisites.html) for details.
+## Setup 
+Before to proceed to build the receiver, you must have installed the following:
+1. Be sure to have installed [nodejs](https://nodejs.org/en/) and [npm](https://www.npmjs.com/)
+2. Optionally, you would like to install [yarn](https://yarnpkg.com/en/)
+3. Once you have installed the above go to the root of the project and install all de dependencies from the command line as follows:
```
Comment (aeng7) [inline README.md]: all de dependencies -> all the dependencies

File: config.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+    "playerBrandingId" : "e48c9b51282d406f957c666149763424",
```
Comment (aldigun) [inline config.json]: Whose playerID is this? This is a sample receiver, therefore customer need to replace this with their own playerID.

Perhaps we need to put a comment here?

File: README.md [hunk 3]
```diff
@@ -1,41 +1,79 @@
-# chromecast-sample-receiver
-This is the repository that contains an open-source Ooyala Chromecast sample receiver. This application should not be used for testing or evaluating Ooyala capabilities.
+# Chromecast Receiver + OOyala Player v4
```
Comment (aeng7) [inline README.md]: Should we be capitalizing the 2nd O in Ooyala?

File: README.md [hunk 4]
```diff
@@ -1,41 +1,79 @@
...
+This repository contains the implementation of Google cast sdk as a receiver for video playback through Chromecast devices. It has an adaptation of OOyala Player v4.
```
Comment (aeng7) [inline README.md]: Should we be capitalizing the 2nd O in Ooyala?

File: dist/cast.js [hunk 5]
```diff
@@ -0,0 +1,2 @@
...ocument)throw new Error("The style-loader cannot be used in a non-browser environment");(e=e||{}).attrs="object"==typeof e.attrs?e.attrs:{},e.singleton||(e.singleton=p()),e.insertInto||(e.insertInto="head"),e.insertAt||(e.insertAt="bottom");var n=o(t,e);return r(n,e),function(t){for(var i=[],a=0;a<n.length;a++){var s=n[a];(u=f[s.id]).refs--,i.push(u)}for(t&&r(o(t,e),e),a=0;a<i.length;a++){var u=i[a];if(0===u.refs){for(var c=0;c<u.parts.length;c++)u.parts[c]();delete f[u.id]}}}};var g=function(){var t=[];return function(e,n){return t[e]=n,t.filter(Boolean).join("\n")}}()},function(t,e){t.exports=function(t){var e="undefined"!=typeof window&&window.location;if(!e)throw new Error("fixUrls requires window.location");if(!t||"string"!=typeof t)return t;var n=e.protocol+"//"+e.host,r=n+e.pathname.replace(/\/[^\/]*$/,"/");return t.replace(/url\s*\(((?:[^)(]|\((?:[^)(]+|\([^)(]*\))*\))*)\)/gi,function(t,e){var o=e.trim().replace(/^"(.*)"$/,function(t,e){return e}).replace(/^'(.*)'$/,function(t,e){return e});if(/^(#|data:|http:\/\/|https:\/\/|file:\/\/\/)/i.test(o))return t;var i;return i=0===o.indexOf("//")?o:0===o.indexOf("/")?n+o:r+o.replace(/^\.\//,""),"url("+JSON.stringify(i)+")"})}}]);
```
Comment (aldigun) [inline dist/cast.js]: Where is the source of this file? Is this Google's? Is this ours?
If this is Google's should we use google's library URL for it?

File: README.md [hunk 6]
```diff
@@ -1,41 +1,79 @@
...
+- warn (*default*)
+- error
+
+You are free to add more logs than the provided on the code by using the logger implementation as follows:
+```javascript
+logger.info("Hi, this is a ingo log")
+logger.error("I'm an error message")
+```
+
+the logger works like de `console.log` statement, so you can use it with the same behaviour.
```
Comment (aeng7) [inline README.md]: de -> the

File: config.json (hunk 2 above)
Comment (aeng7) [inline config.json]: Is this an actual playerBrandingId? Do we want to include an actual playerBrandingId as default here?

File: dist/cast.js (hunk 5 above)
Comment (aldigun) [reply dist/cast.js]: If this is ours.. How are we going to update it if we need to fix this? How is customer going to integrate this? Essentially, should this file be inside valhalla or evergreen production/latest production routes of V4?

@spaddon

File: dist/receiver.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>Ooyala Receiver</title>
+
+    <script type="text/javascript" src="//www.gstatic.com/cast/sdk/libs/receiver/2.0.0/cast_receiver.js"></script>
+    <script type="text/javascript" src="//player.ooyala.com/core/e48c9b51282d406f957c666149763424?plugins=bm"></script>
```
Comment (aeng7) [inline dist/receiver.html]: Do we want to include an actual playerBrandingId here? Also plugins=bm loads only the Bitmovin plugin. Do we need the Bitmovin plugin here?

File: dist/cast.js (hunk 5 above)
Comment (marcosmx) [reply dist/cast.js]: This is generated by webpack as a result of bundling all the code at src folder

File: src/app/idleTimer.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * if the devices was not used for some time and shutdown the app session
+ * 
+ * @module uiManager
+ */
+import * as logger from 'loglevel'
+import castManager from './receiverManager'
+
+const timeout = {
+    PAUSED: 1000 * 60 * 5, // 5 minutes
+    IDLE: 1000 * 60 * 10 // 5 minutes
```
Comment (aeng7) [inline src/app/idleTimer.js]: Comment should be 10 minutes

File: dist/receiver.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    <script type="text/javascript" src="//player.ooyala.com/core/e48c9b51282d406f957c666149763424?plugins=bm"></script>    
+</head>
+<body>
+    <div id="splash-screen" class="fbox-container fs">
+        <img id="splash-image" src="images/logo.svg" />
+        <div id="status-cast">
+        </div>
+    </div>
+    <div id="player"></div>
+<script type="text/javascript" src="cast.js?e9d20cbdcc1904305920"></script></body>
```
Comment (aldigun) [inline dist/receiver.html]: What is this "?e9d20cbdcc1904305920" used for?
This is a follow up on the question above about cast.js. How customer is going to use this? How to fix and update to ALL customer receiver without them changing their page?

File: src/app/player.js [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+     * as a JSON object or string, if the typeof is string a conversion to Object will be
+     * perfomed
+     * 
+     * The params object will be merged with the default values for the Ooyala player
+     *
+     * @param {object} data custom data from sender
+     * @memberof CastPlayer
+     */
+    setAsset(data) {
+        this.ec = data.ec || null;
```
Comment (aeng7) [inline src/app/player.js]: Do we need a null check for data?

File: src/app/player.js [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+    
+    /**
+     * getMessageHandler handles the incoming events at the created message bus and responds for each type
+     * of incoming actions on the message request
+     * 
+     * @param {cast.receiver.CastMessageBus.Event} e 
+     * @memberof CastPlayer
+     */
+    getMessageHandler(e) {
+        logger.info(LOG_PREFIX,"MessageBus Action:", e.data.action);
```
Comment (aeng7) [inline src/app/player.js]: Should we add null checks for e and e.data?

File: src/app/player.js [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+        switch (e.data.action) {
+            case "setCCLanguage":
+                this.setClosedCaptions(e.data.data);
+                break;
+            case "getstatus":
```
Comment (aeng7) [inline src/app/player.js]: Just double checking, is the first s in status supposed to be lower case?

File: src/app/player.js [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+    }
+
+    
+    /**
+     * play executes the play method on the Ooyala player
+     * 
+     * @memberof CastPlayer
+     */
+    play() {
+        this.OOPlayer.play();
```
Comment (aeng7) [inline src/app/player.js]: Should we add a null check for this.OOPlayer?

File: src/app/player.js [hunk 14]
```diff
@@ -0,0 +1,471 @@
...
+    }
+
+    /**
+     * pause executes the pause method on the Ooyala player
+     * 
+     * @memberof CastPlayer
+     */
+    pause() {
+        this.OOPlayer.pause();
```
Comment (aeng7) [inline src/app/player.js]: Should we add a null check for this.OOPlayer?

File: src/app/player.js [hunk 15]
```diff
@@ -0,0 +1,471 @@
...
+
+    
+    /**
+     * seek sets the current position of the player on the asset from the beginning 
+     * 
+     * @param {number} time 
+     * @memberof CastPlayer
+     */
+    seek(time) {
+        this.OOPlayer.seek(time);
```
Comment (aeng7) [inline src/app/player.js]: Should we add a null check for this.OOPlayer?

File: src/app/player.js [hunk 16]
```diff
@@ -0,0 +1,471 @@
...
+
+    /**
+     * setVolume sets the current volume of the Ooyala player. The value must be betweeen
+     * 0 and 1, inclusive
+     * 
+     * @param {any} volume 
+     * @memberof CastPlayer
+     */
+    setVolume(volume) {
+        this.OOPlayer.setVolume(volume.level);
```
Comment (aeng7) [inline src/app/player.js]: Should we add a null check for this.OOPlayer? Should we also use max/min functions to ensure the volume is between 0 and 1? We should also add a null check for the volume parameter.

File: src/app/player.js [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+     * getVolume returns the current volume value of the Ooyala player. The value is converted
+     * to be an instance of cast.receiver.media.Volume
+     * 
+     * @returns {cast.receiver.media.Volume}
+     * @memberof CastPlayer
+     */
+    getVolume() {
+        var volume = new cast.receiver.media.Volume();
+        volume.level = this.OOPlayer.getVolume();
+        volume.muted = false;
```
Comment (aeng7) [inline src/app/player.js]: Do we not use this muted state? Why is this always false?

File: src/app/player.js [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+    /**
+     * getVolume returns the current volume value of the Ooyala player. The value is converted
+     * to be an instance of cast.receiver.media.Volume
+     * 
+     * @returns {cast.receiver.media.Volume}
+     * @memberof CastPlayer
+     */
+    getVolume() {
+        var volume = new cast.receiver.media.Volume();
+        volume.level = this.OOPlayer.getVolume();
```
Comment (aeng7) [inline src/app/player.js]: Null check for this.OOPlayer

File: src/app/player.js [hunk 19]
```diff
@@ -0,0 +1,471 @@
...
+
+    /**
+     * onCreateHandlers receives the instance of the Ooyala player and create event listeners
+     * for the Ooyala player events
+     * 
+     * @param {OO.Player} player 
+     * @memberof CastPlayer
+     */
+    onCreateHandlers (player) {  
+        player.mb.subscribe(OO.EVENTS.PLAYBACK_READY, PLAYERNAMESPACE, this.onPlaybackReady.bind(this));
```
Comment (aeng7) [inline src/app/player.js]: Null check for the player parameter.

File: src/app/player.js [hunk 20]
```diff
@@ -0,0 +1,471 @@
...
+     * the senders about the event
+     * 
+     * @param {string} event 
+     * @param {number} time 
+     * @memberof CastPlayer
+     */
+    onSeeked(event, time) {
+        // if the player is in PAUSE state then try to update the scrubber bar to the
+        // actual time
+        var skin = this.OOPlayer.modules.find((m) => {return m.name == "Html5Skin";});
```
Comment (aeng7) [inline src/app/player.js]: Is there something we can do to avoid this dependency on the skin?

File: src/app/receiverManager.js [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+import UIManager from './uiManager'
+import Timer from './idleTimer'
+
+const LOG_PREFIX = 'ReceiverManager:';
+
+var castReceiverManager = cast.receiver.CastReceiverManager.getInstance();
+
+castReceiverManager.onSenderConnected = onSenderConnected;
+castReceiverManager.onSenderDisconnected = onSenderDisconnected;
+castReceiverManager.onReady = onReady;
```
Comment (aeng7) [inline src/app/receiverManager.js]: Where do onSenderConnected, onSenderDisconnected, and onReady come from?

File: dist/receiver.html (hunk 9 above)
Comment (aldigun) [reply dist/receiver.html]: Based on our discussion, 
1. we will need to point this to a static PATH: example:
//player.ooyala.com/static/chromecast/LOCATION_OF_CORRECT_CAST.JS
This cast.js need to be proper location on candidate or sandbox or latest (production)

2. //player.ooyala.com/core/e48c9b51282d406f957c666149763424?plugins=bm
This need to be //player.ooyala.com/core/e48c9b51282d406f957c666149763424?plugins=bm,cast
Assuming that valhalla ignores non-existent plugin, we can placehold "cast" there for future update

3. We need to migrate cast.js into core (mjolnir) as part of V4 usual deployment

4. Once player make production deployment, have PBS map "cast" into proper V4 "cast" plugin.

File: dist/receiver.html (hunk 7 above)
Comment (aldigun) [inline dist/receiver.html]: How is e48c9b51282d406f957c666149763424 used? is this hard-coded or generated? How customer should replace this with their playerID?

**Pull request merged**